cvmfs/sync_union_tarball.cc

```diff
@@ -4,30 +4,223 @@
+  assert(ARCHIVE_OK == archive_read_support_format_empty(src));
+
+  if (tarball_path_ == "--") {
+    result = archive_read_open_filename(src, NULL, kBlockSize);
+  } else {
+    std::string tarball_absolute_path = GetAbsolutePath(tarball_path_);
+    result = archive_read_open_filename(src, tarball_absolute_path.c_str(),
+                                        kBlockSize);
+  }
+
+  if (result != ARCHIVE_OK) {
+    LogCvmfs(kLogUnionFs, kLogStderr, "Impossible to open the archive.");
+    return false;
+  }
+
+  return SyncUnion::Initialize();
+}
+
 /*
- * Still have to underastand how the madiator, scratch, read and union path
- * works together.
+ * Libarchive is not thread aware, so we need to make sure that before
+ * to read/"open" the next header in the archive the content of the
+ * present header is been consumed completely.
+ * Different thread read/"open" the header from the one that consumes
+ * it so we opted for a Signal that is backed by a conditional variable.
+ * We wait for the signal just before to read the header.
+ * Then when we have done with the header the Signal is fired.
+ * The Signal can be fired inside the main loop if we don't need to read
+ * data, or when the IngestionSource get closed, which means that we are
+ * not reading data anymore from there.
+ * This whole process is not necessary for directories since we don't
+ * actually need to read data from them.
  */
-SyncUnionTarball::SlyncUnionTarball(SyncMediator *mediator,
-                                    const std::string &tarball_path,
-                                    const std::string &base_directory)
-    : tarball_path_(tarball_path), base_directory_(base_directory) {}
+void SyncUnionTarball::Traverse() {
+  assert(this->IsInitialized());
+  struct archive_entry *entry = archive_entry_new();
 
-/*
- * We traferse the tar and then we keep track (the path inside a set) of each
- * .tar we find, then we go back and repeat the procedure. Should we be worry by
- * tar-bombs? Maybe, we will think about it later.
+  /*
+   * As first step we eliminate the directories we are request.
+   */
+  if (to_delete_ != "") {
+    vector<std::string> to_eliminate_vec = SplitString(to_delete_, ':');
+
+    for (vector<string>::iterator s = to_eliminate_vec.begin();
+         s != to_eliminate_vec.end(); ++s) {
+      std::string parent_path;
+      std::string filename;
+      SplitPath(*s, &parent_path, &filename);
+      if (parent_path == ".") parent_path = "";
+      SharedPtr<SyncItem> sync_entry =
+          CreateSyncItem(parent_path, filename, kItemDir);
+      mediator_->Remove(sync_entry);
+    }
+  }
+
+  while (true) {
+    /* Get the lock, wait if lock is not available yet */
+    read_archive_signal_->Wait();
+
+    int result = archive_read_next_header2(src, entry);
+
+    switch (result) {
+      case ARCHIVE_FATAL: {
+        LogCvmfs(kLogUnionFs, kLogStderr,
+                 "Fatal error in reading the archive.");
+        return;
```

Comment:
This needs to be an `abort()`, I think. Otherwise we might publish half an archive.